pkg/cloud/rgraph/exec/retrieable_action.go

```diff
@@ -0,0 +1,51 @@
+/*
+Copyright 2024 Google LLC
+
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package exec
+
+import (
+	"context"
+
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud"
+)
+
+// retriableAction is a decorator to the action adding retry mechanism
+type retriableAction struct {
+	Action
+	isRetriable func(error) bool
+}
+
+// NewRetriableAction decorates Action with retry provider function
+func NewRetriableAction(a Action, rp func(error) bool) Action {
+	return &retriableAction{a, rp}
+}
+
+// Run executes Action. If error is returned retry provider checks if Action
+// should be rerun
+func (ra *retriableAction) Run(ctx context.Context, c cloud.Cloud) (EventList, error) {
+	for {
+		events, err := ra.Action.Run(ctx, c)
```

Comment:
retry policy is implemented by rate limiter. Every time you call GCE API it is going through Rate limiter.

Reply:
if needed it can be also implemented in retry provider

Reply:
Retry Provider is an simple function as of now.

Can you point me to the code where retry policy is implemented within rate limiter? 

Do we retry actions in rate limiter when we are getting quota errors (I also couldn't find it).

Can't we have retry policy in both retriable action and rate limiter? Sometimes from rate limiter perspective we do not have to wait longer (as we still have capacity to call GCE API), but from an action perspective it maybe have some sense to wait before retrying.

Reply:
good point, retry action which will be used is not implemented yet. We can add there back off logic

Reply:
There are standard backoff retry packages. I don't think it makes sense to push the rate limiting of retries down to the overall rate limit. What that will do is that if a operation is fast and retries right away, it will exhaust the rate limiter below -- potentially starving other operations.

The delay from the retry should be exposed here to avoid hiding that from the user.

Reply:
I agree with that. This is inline with what I wrote and I think Katarzyna is already working on fixing that.